internal/grpc/handler.go

```diff
@@ -0,0 +1,152 @@
+	"github.com/golang/protobuf/protoc-gen-go/descriptor"
+	"github.com/uber/prototool/internal/desc"
+	"github.com/uber/prototool/internal/extract"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/keepalive"
+)
+
+type handler struct {
+	logger         *zap.Logger
+	callTimeout    time.Duration
+	connectTimeout time.Duration
+	keepaliveTime  time.Duration
+	headers        []string
+
+	getter extract.Getter
+}
+
+func newHandler(options ...HandlerOption) *handler {
+	handler := &handler{
+		logger: zap.NewNop(),
+	}
+	for _, option := range options {
+		option(handler)
+	}
+	if handler.callTimeout == 0 {
+		handler.callTimeout = DefaultCallTimeout
+	}
+	if handler.connectTimeout == 0 {
+		handler.connectTimeout = DefaultConnectTimeout
+	}
+	// TODO(pedge): composition
+	handler.getter = extract.NewGetter(
+		extract.GetterWithLogger(handler.logger),
+	)
+	return handler
+}
+
+func (h *handler) Invoke(fileDescriptorSets []*descriptor.FileDescriptorSet, address string, method string, inputReader io.Reader, outputWriter io.Writer) error {
+	descriptorSource, err := h.getDescriptorSourceForMethod(fileDescriptorSets, method)
+	if err != nil {
+		return err
+	}
+	clientConn, err := h.dial(address)
+	if err != nil {
+		return err
+	}
+	defer func() { _ = clientConn.Close() }()
+	invocationEventHandler := newInvocationEventHandler(outputWriter, h.logger)
+	ctx, cancel := context.WithTimeout(context.Background(), h.callTimeout)
+	defer cancel()
+	if err := grpcurl.InvokeRpc(
+		ctx,
+		descriptorSource,
+		clientConn,
+		method,
+		h.headers,
+		invocationEventHandler,
+		decodeFunc(inputReader),
+	); err != nil {
+		return err
+	}
+	return invocationEventHandler.Err()
+}
+
+func (h *handler) dial(address string) (*grpc.ClientConn, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), h.connectTimeout)
+	defer cancel()
+	return grpcurl.BlockingDial(ctx, "tcp", address, nil, h.getDialOptions()...)
+}
+
+func (h *handler) getDialOptions() []grpc.DialOption {
+	var dialOptions []grpc.DialOption
+	if h.keepaliveTime != 0 {
+		dialOptions = append(
+			dialOptions,
+			grpc.WithKeepaliveParams(
+				keepalive.ClientParameters{
+					Time:    h.keepaliveTime,
+					Timeout: h.keepaliveTime,
+				},
+			),
+		)
+	}
+	return dialOptions
+}
+
+func (h *handler) getDescriptorSourceForMethod(fileDescriptorSets []*descriptor.FileDescriptorSet, method string) (grpcurl.DescriptorSource, error) {
+	servicePath, err := getServiceForMethod(method)
+	if err != nil {
+		return nil, err
+	}
+	service, err := h.getter.GetService(fileDescriptorSets, servicePath)
+	if err != nil {
+		return nil, err
+	}
+	fileDescriptorSet, err := desc.SortFileDescriptorSet(service.FileDescriptorSet, service.FileDescriptorProto)
+	if err != nil {
+		return nil, err
+	}
+	return grpcurl.DescriptorSourceFromFileDescriptorSet(fileDescriptorSet)
+}
+
+func getServiceForMethod(method string) (string, error) {
+	split := strings.Split(method, "/")
```

Comment:
Is `package.service/Method` a gRPC standard or just a `grpcurl` standard? If it's just `grpcurl` how would you feel if we separated these into two fields instead? `... package.service Method ...`

Reply:
It’s a gRPC standard, the request url is /package.service/method